### PR TITLE
Refactor the use of pattern matcher in ptx

### PR DESCRIPTION
**Pull request by SzymonOzog**
Updated as promised. I left out `ptr_arm` as a separate function as it's quite complex and needs to be called for all LOAD/STORE not fitting that well with the pattern matcher

### CODE REVIEW

File: tinygrad/renderer/assembly.py [hunk 1]
```diff
@@ -106,6 +85,18 @@ def ptr_ar(root):
       if o in u.vin and u is not n:
         u.vin = tuple(n if x == o else x for x in u.vin)
     if rew := matcher.rewrite(u): replace[u] = rew
+
+  for o,n in replace.items():
```
Comment (geohot) [inline tinygrad/renderer/assembly.py:89]: What does this do? Should this be in the pattern matcher?

Comment (SzymonOzog) [reply tinygrad/renderer/assembly.py:89]: This goes through the uop returned by the fxn passed into the pattern matcher and all of it's parents and adds them to the UOpGraph if they aren't there

Comment (geohot) [reply tinygrad/renderer/assembly.py:89]: That should be in pattern matcher!

Comment (SzymonOzog) [reply tinygrad/renderer/assembly.py:89]: oh it works on `replace` continuing on the work done in the previous block - not sure how to fit it into the pattern matcher, could you elaborate on that?

**Comment by geohot** [comment]
Meh, I'll merge it, it's less lines. But this should really be generic code to apply a pattern to a formed list of uops.

**Pull request merged**